data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/OpenSearchClientRefresher.java

```diff
@@ -0,0 +1,52 @@
+package org.opensearch.dataprepper.plugins.source.opensearch;
+
+import org.opensearch.client.opensearch.OpenSearchClient;
+import org.opensearch.dataprepper.model.plugin.PluginComponentRefresher;
+import org.opensearch.dataprepper.plugins.source.opensearch.worker.client.OpenSearchClientFactory;
+
+import java.util.Objects;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class OpenSearchClientRefresher
+        implements PluginComponentRefresher<OpenSearchClient, OpenSearchSourceConfiguration> {
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    private final OpenSearchClientFactory openSearchClientFactory;
+    private OpenSearchSourceConfiguration existingConfig;
+    private OpenSearchClient currentClient;
+
+    public OpenSearchClientRefresher(final OpenSearchClientFactory openSearchClientFactory,
+                                     final OpenSearchSourceConfiguration openSearchSourceConfiguration) {
+        this.openSearchClientFactory = openSearchClientFactory;
+        existingConfig = openSearchSourceConfiguration;
+        currentClient = openSearchClientFactory.provideOpenSearchClient(openSearchSourceConfiguration);
+    }
+
+    @Override
+    public OpenSearchClient get() {
+        readWriteLock.readLock().lock();
+        try {
+            return currentClient;
+        } finally {
+            readWriteLock.readLock().unlock();
+        }
+    }
+
+    @Override
+    public void update(final OpenSearchSourceConfiguration openSearchSourceConfiguration) {
+        if (basicAuthChanged(openSearchSourceConfiguration)) {
+            readWriteLock.writeLock().lock();
+            try {
+                currentClient = openSearchClientFactory.provideOpenSearchClient(openSearchSourceConfiguration);
```

Comment:
This class can be consolidated with `ElasticsearchClientRefresher` by using generics and a `Function`.

This line appears to be the only difference. So you can use a `Function<OpenSearchSourceConfiguration, C>`. This line then becomes:

```
currentClient = clientFunction.apply(openSearchSourceConfiguration);
```

Reply:
This has type erasure issue in `OpenSearchIndexPartitionCreationSupplier`

```
final PluginComponentRefresher clientRefresher = clusterClientFactory.getClientRefresher();

        if (clientRefresher instanceof OpenSearchClientRefresher) {
            this.openSearchClientRefresher = (OpenSearchClientRefresher) clientRefresher;
        } else if (clientRefresher instanceof ElasticsearchClientRefresher) {
            this.elasticsearchClientRefresher = (ElasticsearchClientRefresher) clientRefresher;
        } else {
            throw new IllegalArgumentException(String.format("ClusterClientFactory provided an invalid client object to the index partition creation supplier. " +
                    "The clientRefresher must be of type OpenSearchClientRefresher. The clientRefresher passed is of class %s", clientRefresher.getClass()));
        }
```

Reply:
There is a common solution here which is to provide the class in the constructor.

```
public OpenSearchClientRefresher(Class<C> clientClass, Function<OpenSearchSourceConfiguration, C> clientGenerationFunction) {
  ...
}
```

Then, make a getter to get the Class:

```
Class<C> getClientClass() {
  return clientClass;
}
```

Now, you can query this using `isAssignableFrom`.

```
if (OpenSearchClient.class.isAssignableFrom(clientRefresher.getClientClass()) {
  this.openSearchClientRefresher = clientRefresher;
} else if (ElasticsearchClient.class.isAssignableFrom(clientRefresher.getClientClass()) {
  this.elasticsearchClientRefresher = clientRefresher;
} else {
  ...
}
```